Stackmap updates to the withfield and defaultvalue instructions

XxAdi101xX opened the pull request:
This commit does the following:
- updates the erroraneous values associated with the defaultvalue bytecode instruction
- adds the missing withfield switchcase in stackmap.c
- fixes some bugs regarding the withfield instruction in StackMap.java

Signed-off-by: Adithya Venkatarao <adi_101@live.com>

XxAdi101xX commented:
@tajila ready for review

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -137,6 +137,23 @@
 	static double defaultDoubleNew = -123412341.21341234d;
 	static float defaultFloatNew = -123423.12341234f;
 	static Object defaultObjectNew = (Object)0xFFEEFFEE;
+	static class CircularBuffer {
```
add a type parameter

tajila replied:
this should be moved out somewhere it can be reused again

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -137,6 +137,23 @@
+		private int size;
+		private int head;
+		
+		public Object array[];
```
This should be defined as a type parameter

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1491,6 +1508,21 @@ static public void testDefaultValueInassortedRefWithLongAlignmentArray() throws
+	 * This forces the GC to unload the classes.
+	 */
+	@Test(priority=4)
+	static public void testPersistantFlattenedValueTypeArray() throws Throwable {
```
add tests that create an array of objectwith[single|ref|double] alignment

XxAdi101xX commented:
@tajila only the VM changes have been added to this PR, please review

XxAdi101xX commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/PCStack.java:
```diff
@@ -1017,7 +1017,7 @@
 		0x00 /* JBbreakpoint = 202  -- pops: 0 pushes: 0*/ ,
-		0x01 /* JBdefaultvalue = 203 -- pops: 0 pushes: 1*/ ,
+		0x50 /* JBdefaultvalue = 203 -- pops: 0 pushes: 1*/ ,
 		0x80 /* JBwithfield = 204  -- pops: 0 pushes: 0*/ ,
```
Withfield seems to overlap with 0x80 `/* JBaload (16r19) <oslot><read>*/` and defaultvalue overlaps with `0x50 /* JBfstore (16r38) <islot><write>*/`. Moreover, there are existing values of `JBdefaultvalue (16rCB)` and `JBwithfield (16rCC)` defaulted to 0x0. Can the hex for the unimplemented bytecodes just be altered to prevent overlap or is there a standard that we follow regarding the association of an instruction with a bytecode?

tajila replied:
It looks like there are some extra unimplemented bytecodes. This code should match the VM version in pcstack.c

XxAdi101xX replied:
I've compared the code for `defaultvalue` and `withfield` in both `pcstack.c` and `PCstack.java` and they are the same actually

tajila commented:
> changes have been made; regarding the unimplemented bytecode instruction comment, as mentioned above, I've compared the code in pcstack.c and PCstack.java for defaultvalue and withfield and they seem to be the same, hence I've not made any additional changes to that

```
		0x0 /* JBunimplemented (16rCB) */,
		0x0 /* JBunimplemented (16rCC) */,
```

Those unimplemented bytecodes do not exist in the VM version

XxAdi101xX commented:
@tajila I've updated `pcstack.c` to mirror `PCstack.java` addressed the other comments, and added indentation changes to maintain stylistic consistency for arrays

tajila commented:
> I've updated pcstack.c to mirror PCstack.java addressed the other comments, and added changes to maintain stylistic consistency for arrays

Dont about whitespace differences in pcstack, just the comments

tajila commented on runtime/stackmap/stackmap.c:
```diff
@@ -436,7 +436,7 @@ mapStack(UDATA *scratch, UDATA totalStack, U_8 * map, J9ROMClass * romClass, J9R
 													((J9ROMFieldRef *) (&(pool[index]))));
 				signature = (U_8) J9UTF8_DATA(utf8Signature)[0];
-				if ((signature == 'L') || (signature == '[')) {
+				if ((signature == 'L') || (signature == '[') || (signature == 'Q')) {
```
nitpick: compare 'L' then 'Q' then '['

XxAdi101xX commented:
@tajila addressed comment and removed duplicate bytecode that were causing build failures, please review again

XxAdi101xX commented:
@tajila reverted whitespace changes

tajila commented:
@gacholio please review these changes

gacholio commented on runtime/util/argcount.c:
```diff
@@ -20,19 +20,38 @@
+    0, /* Y */
+    1, /* Z -> boolean */
+    1  /* [ -> 1-D array reference */
+}; 
```
The array size here has changed from 28 to 27. The stripped character is /, so this should not matter, but I'd like it verified that the change was intentional.

XxAdi101xX replied:
Yes, I intentionally removed it because it was a character that isn't currently used to represent a java field descriptor.

gacholio commented:
jenkins test sanity xlinux jdk11

XxAdi101xX commented:
@gacholio pls hold off on merging temporarily, my personal builds were failing on a jitcompiler issue so I've rebased on another branch with this commit cherry picked and I'm rerunning the tests on both linux and window

XxAdi101xX commented:
@gacholio builds are passing, the pr is safe to merge
